### ISSUE TITLE
DUP/PRDT : Explore notification service options

### ISSUE THREAD

**Issue by meyerdarcie** [bcgov/parks-reso-public#398]
#### Description:
This ticket is to look at notification service options, such as AWS SES, that are robust, flexible and nimble. 

The notification service should be robust enough to:
- Meet the demand of sending text messages to DUP users on busy weekends 
- Support sending facility specific, custom reminder emails 
- Alerts backcountry recreators who want to be notified of availability for certain experiences
- Provide frontcountry campers who request a reminder text

The outcome of this ticket is to define how we are actually going to build the new notification service.

#### Acceptance Criteria:
- [ ] Privacy implications are included, will be used for PIA/STRA
- [ ] Cost implication are noted
- [ ] Define architecture

#### Development Checklist:

- [ ]

#### Dependencies

- Blocked by
- Blocking

**Relevant documentation as reference**

**Definition of Ready**

- [ ] Acceptance criteria are included
- [ ] Wireframes are included (if applicable)
- [ ] Design / Solution is accepted by Product Owner (if applicable)
- [ ] Dependencies are identified (technical, business, regulatory/policy)
- [ ] Story has been estimated (under 13 pts)

**Definition of Done** 
- In progress:
  - [ ] Acceptance criteria are tested (Functionality meets the acceptance criteria defined in the ticket)
  - [ ] UI meets accessibility requirements
  - [ ] Unit tests are written
  - [ ] Work is traceable in GitHub
  - [ ] PR linked to ticket number
  - [ ] If needed/required - Dev adds flag/label to highlight any migration steps necessary prior to PROD deployment
- Code review:
  - [ ] Code is peer reviewed and has passed CI/CD tests
- QA:
  - [ ] Acceptance criteria are tested (Functionality meets the acceptance criteria defined in the ticket)
  - [ ] Code is potentially shippable to the production environment
  - [ ] Functional features have been tested and passed by QA
  - [ ] UI components tested by designer
  - [ ] Code is deployed to PROD when moved to 'done' column (unless requested otherwise by PO)
- PO Review:
  - [ ] Acceptance criteria are tested (Functionality meets the acceptance criteria defined in the ticket)
  - [ ] Reviewed and approved by Product Owner

#### Notes:
- Leverage Amazon SES for sending emails
  - unlock added email sending limits by request
- Leverage Amazon SNS for sending TXTs
- Handlebars/Moustache (or other npm library) for HTML templates in order to send plain and html content to email recipients

**Comment by Dianadec** [comment]
Notes from June 11 Refinement: 
We'll have to move away from gcnotify to AWS SES eventually

**Comment by Dianadec** [comment]
@marklise please add anything to the dev checklist as needed

**Comment by Dianadec** [comment]
Refinement notes:
- Some things can be customized per park, per facility
- Is the functionality of offering text messages as an option for people built in a way that we can reduce the number of facilities? (could be implemented - cam)
- Turn off sms for certain facilitates?? b/c if we were to remove this as an option for the beach access at Golden Ears (most ample parking) - we would no longer run into this limit. The whole problem is the beach access - there's over 1000 alone. At Joffre there's 500 passes (but actually prb only 180 passes that are getting email/texts. Garibaldi/golden ears are quite limited.
- We have that winter warning flag, this would be akin to that

**Comment by Dianadec** [comment]
TO BE CONFIRMED - different service? or build new features to support current gc notify service. @meyerdarcie 

Idea: Update ticket with info:
- Turn off Main Parking - Beach access @ Golden Ears "Allouette Lake South Beach Day Use Parking Lot"

Note: This is connected to public/admin portion too - need to have some design changes - can't give people an option to sign up for text notifications

admin side - whether they want text to be sent to a park

... or, build a feature for the admin side for them to update it themselves

UX perspective: Will need to look at text to make it clear to visitors that certain facilities don't have text message notifications. - need some kind of dislaimer/notification that explains why they see it sometimes and sometimes not.

**Comment by Christopher-walsh22** [comment]
Through researching this ticket I have come up with a few suggestions on services we would want to use and considerations for the final decision. 

The architecture I recommend using combines two AWS services: AWS SES, and AWS End-User Messaging (Formerly Amazon Pinpoint). SES would be used to handle email notifications, including confirmation messaging and reminders. While End-User Messaging would be used for SMS reminders depending on the facility and whether users opt in. 

Originally proposed in this ticket was to use SNS for sending text messages. I don't believe this is the correct approach as we this service does not support sending attachments should we require them in the future. SNS is also designed to notify multiple subscribers with the same information. While we could potentially adapt SNS to send personalized text messages, it is not designed for this purpose. 

SES will check all the boxes for our email requirements. Costing $0.10/1000 emails, with an additional charge of $0.12 per GB of attachments. 

End-User Messaging will have some additional costs to set up due to requiring a phone number to send the text messages from. Pricing tables on this become available during the set up phase. The price to send each message would depend on the type of phone number registered to bc parks (selected during set up process), Using a 10digit phone number the cost to send a single SMS $0.01348. End-User Messaging is also used by the Provincial Health Services Authority to handle booking confirmation, notification, and vaccine reminders to residents of British Columbia. 

As managed AWS services, both End-User Messaging and SES are protected by AWS’s global security procedures, ensuring that all data passed through these services is handled securely and in accordance with their privacy policies.

**Comment by Dianadec** [comment]
@marklise  for your review please

**Comment by marklise** [comment]
SNS is fine for simple `reminder` type notifications, however as stated not suitable for attachments or anything other than basic text.  The suggestion to go with SES on emailing out is good as is End-User Messaging for TXT.  For email there are several templating modules to use in order to customize the content.

**Comment by LindsayMacfarlane** [comment]
Thanks so much for this research and recommendation @Christopher-walsh22 ! Curious on next steps for exploring this further?

- You mentioned that the Health Authority uses this approach. Have we reached out to anyone to see if they like it from an administrative perspective?
- Do we need a spike ticket to look more at the implementation?
- Assuming we will need to add this to our PIA ;)

**Comment by LindsayMacfarlane** [comment]
Follow up with Chris, December 18, 2024: 

- Have not reached out to anyone from Health. Testimonial on amazon about their use.
- Next steps: spike ticket to implement the service and test out. You need accounts to register/use the service. We may want to create 
- This will replace GC notify
- GC notify is fine when it works, but when we re-tool or change things, it can be quite problematic/challenging.
- Confirm: do we get charged for GC notify?
- Would need a phone number (short vs long) to send the message from

**Comment by LindsayMacfarlane** [comment]
New ticket created to capture next steps: https://github.com/orgs/bcgov/projects/49/views/2?pane=issue&itemId=91421606&issue=bcgov%7Creserve-rec-public%7C86

Thanks for the quick call to work through things @Christopher-walsh22 ! Much appreciated :)
